build: add lint rule to enforce functions are moved to outermost scopes

Planeshifter opened the pull request:
## Description

> What is the purpose of this pull request?

This pull request:

-  adds a new ESLint rule flagging function declarations that are unnecessarily nested and can be moved to a higher scope, helping avoid repeated function creation and improving performance, while respecting intentional patterns like closures and IIFEs.

Rationale: Functions defined inside other functions are recreated on each invocation, causing unnecessary memory allocation and performance overhead. Moving independent functions to module scope ensures they are defined only once while preserving legitimate use cases like closures that access outer variables.

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves https://github.com/stdlib-js/metr-issue-tracker/issues/65

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

Planeshifter committed:
build: add lint rule to enforce functions are moved to outermost possible scope

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte committed:
docs: move variable declarations

Signed-off-by: Athan <kgryte@gmail.com>

kgryte committed:
docs: move variable declarations

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/_tools/eslint/rules/no-unnecessary-nested-functions/examples/index.js:
```diff
@@ -0,0 +1,61 @@
+var code = [
+	'function outer() {',
+	'  function inner() {',
+	'	return 42;',
```
```suggestion
	'    return 42;',
```

kgryte committed:
style: fix indentation

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/_tools/eslint/rules/no-unnecessary-nested-functions/test/fixtures/invalid.js:
```diff
@@ -0,0 +1,375 @@
+var test;
+
+// Simple nested function without dependencies:
+test = {
```
```suggestion

// Simple nested function without dependencies:
var test = {
```

kgryte committed:
test: move declaration

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/_tools/eslint/rules/no-unnecessary-nested-functions/test/fixtures/unvalidated.js:
```diff
@@ -0,0 +1,40 @@
+var test;
+
+// Function declaration inside IIFE (intentionally isolated):
+test = {
```
```suggestion

// Function declaration inside IIFE (intentionally isolated):
var test = {
```

kgryte committed:
test: move declaration

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/_tools/eslint/rules/no-unnecessary-nested-functions/test/fixtures/valid.js:
```diff
@@ -0,0 +1,231 @@
+var test;
+
+// Top-level function (no nesting):
+test = {
```
```suggestion

// Top-level function (no nesting):
var test = {
```

kgryte committed:
test: move declaration

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/_tools/eslint/rules/no-unnecessary-nested-functions/test/fixtures/valid.js:
```diff
@@ -0,0 +1,230 @@
+	'code': [
+		'function outer() {',
+		'  function helper() {',
+		'    return this.value * 2;',
```
@Planeshifter Why wouldn't this be a case where we can lift to module scope?

kgryte commented on lib/node_modules/@stdlib/_tools/eslint/rules/no-unnecessary-nested-functions/test/fixtures/valid.js:
```diff
@@ -0,0 +1,230 @@
+	'code': [
+		'function outer() {',
+		'  function helper() {',
+		'    return arguments.length;',
```
@Planeshifter Why wouldn't this be a case where we can lift to module scope?

kgryte commented on lib/node_modules/@stdlib/_tools/eslint/rules/no-unnecessary-nested-functions/lib/main.js:
```diff
@@ -0,0 +1,356 @@
+* @private
+* @param {string} varName - variable name
+* @param {Object} startScope - scope to start searching from
+* @returns {Object|null} scope where variable is defined or null if not found
```
```suggestion
* @returns {(Object|null)} scope where variable is defined or null if not found
```

kgryte commented on lib/node_modules/@stdlib/_tools/eslint/rules/no-unnecessary-nested-functions/lib/main.js:
```diff
@@ -0,0 +1,356 @@
+* @private
+* @param {ASTNode} node - function declaration node
+* @param {Object} context - ESLint context
+* @returns {Object|null} highest scope object or null if cannot be moved
```
```suggestion
* @returns {(Object|null)} highest scope object or null if cannot be moved
```

kgryte commented on lib/node_modules/@stdlib/_tools/eslint/rules/no-unnecessary-nested-functions/lib/main.js:
```diff
@@ -0,0 +1,356 @@
+		highestScope = highestScope.upper;
+	}
+
+	// If we're already at module level, we still want to check if we can move there
+	// (This handles the case where the function is directly nested in a top-level function)
```
```suggestion
	// If we're already at module level, we still want to check if we can move there (note: this handles the case where the function is directly nested in a top-level function)...
```

kgryte committed:
Apply suggestions from code review

Signed-off-by: Athan <kgryte@gmail.com>

kgryte reviewed (changes requested):
Other than nested usage of `this` and `arguments`, this is looking fine.

Planeshifter committed:
build: remove special handling for functions accessing `this` and `arguments`

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Planeshifter commented:
Upon reflection, there does indeed appear to be no reason why we couldn't elevate functions accessing `this` or `arguments`. Removed the corresponding logic.

kgryte commented on lib/node_modules/@stdlib/_tools/eslint/rules/no-unnecessary-nested-functions/README.md:
```diff
@@ -0,0 +1,183 @@
+
+<section class="intro">
+
+This rule enforces a convention in stdlib to only have functions defined inside of other functions if they need access to variables defined in the outer function scope. Otherwise, functions should always be elevated to the highest level (ideally module scope) to avoid redefining the functions per invocation.
```
```suggestion
This rule enforces a best practice to only have functions defined inside of other functions if they need access to variables defined in the outer function scope. Otherwise, functions should always be elevated to the highest level (ideally module scope) to avoid redefining the functions per invocation.
```

kgryte replied:
This isn't merely a stdlib convention; this is best practice. You can significantly degrade performance, especially in hot loops, if you are constantly creating new scopes on the stack.

kgryte committed:
docs: update copy

Signed-off-by: Athan <kgryte@gmail.com>

kgryte reviewed (approved):
LGTM
